Resource to manage machine learning jobs

csmarchbanks opened the pull request:
Draft PR to create a terraform resource for creating, updating, and deleting machine learning jobs. The client has not been pushed quite yet while it undergoes a bit of review, but figured we could start reviewing this in the meantime.

csmarchbanks committed:
Resource to manage machine learning jobs

sd2k commented on grafana/resource_machine_learning_job.go:
```diff
@@ -0,0 +1,157 @@
+				Required:    true,
+			},
+			"datasource_type": {
+				Description: "The type of datasource to use when querying. Currently allowed values are prometheus, graphite, loki, postgres, and datadog.",
```
```suggestion
				Description: "The type of datasource being queried. Currently allowed values are prometheus, graphite, loki, postgres, and datadog.",
```

sd2k commented on grafana/resource_machine_learning_job.go:
```diff
@@ -0,0 +1,157 @@
+				Default:     300,
+			},
+			"hyper_params": {
+				Description: "The hyper parameters used to fine tune the algorithm. See https://grafana.com/docs/grafana-cloud/machine-learning/models/ for the full list of available hyper parameters.",
```
For consistency:

```suggestion
				Description: "The hyperparameters used to fine tune the algorithm. See https://grafana.com/docs/grafana-cloud/machine-learning/models/ for the full list of available hyperparameters.",
```

sd2k commented on grafana/resource_machine_learning_job_test.go:
```diff
@@ -0,0 +1,46 @@
+			},
+		},
+	})
+}
```
If it's easy to do, it'd be great to have a failing test case here; something like an invalid datasource type or a job trying to train on too many samples, to show that error messages are useful.

sd2k reviewed (changes requested):
Generally LGTM, just a few small suggestions and test request if it's easy to add.

julienduchesne commented on grafana/resource_machine_learning_job_test.go:
```diff
@@ -0,0 +1,74 @@
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheckCloud(t) },
+		ProviderFactories: testAccProviderFactories,
```
For completeness, you can have: 
- a `CheckDestroy` test that checks that the job is actually destroyed after the test: See example -> `testAccFolderCheckDestroy`
- a custom check to test that the job is actually deployed (not just Terraform blindly setting some attributes): See example -> `testAccFolderCheckExists`

csmarchbanks replied:
Awesome, thanks for pointing those checks out! I have added a second commit implementing both of them.

julienduchesne reviewed (approved):
LGTM! Very nice work! I have one comment but it's not a blocker. Up to you to decide if you want to implement these additional checks. Let me know!

csmarchbanks committed:
Add checks for server ML resources

Add tests to make sure the server has a Job, and make sure that after
deletion the server no longer has a Job.

sd2k reviewed (approved):
LGTM! 👍

csmarchbanks commented:
@julienduchesne All other dependencies have been merged, so this is ready to go now!
